File: pkg/heap/merge_k_lists.go
package heap

import (
	"container/heap"
	. "gocode/types"
)

/**
 * 23. Merge k Sorted Lists
 */
func mergeKLists(lists []*ListNode) *ListNode {
	minHeap := &Heap[*ListNode]{
		less: func(i *ListNode, j *ListNode) bool {
			return i.Val < j.Val
		},
	}

	for _, list := range lists {
		if list != nil {
			heap.Push(minHeap, list)
		}
	}

	dummy := &ListNode{}
	cur := dummy
	for minHeap.Len() > 0 {
		tmp := heap.Pop(minHeap).(*ListNode)
		cur.Next = tmp
		cur = tmp

		if tmp.Next != nil { // push the next item in sorted linked list
			heap.Push(minHeap, tmp.Next)
		}
	}

	return dummy.Next
}

File: python/heap/merge_k_lists.py
import sys, os
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from common import ListNode

from typing import List, Optional
import heapq

"""
23. Merge k Sorted Lists
"""
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        minHeap: List[tuple[int, int, ListNode]] = []
        for i, node in enumerate(lists):
            if node:
                heapq.heappush(minHeap, (node.val, i, node))

        dummy: ListNode = ListNode()
        cur = dummy
        while len(minHeap) > 0:
            _, i, top = heapq.heappop(minHeap)
            cur.next = top
            cur = cur.next
            if top.next:
                heapq.heappush(minHeap, (top.next.val, i, top.next))

        return dummy.next
